packages/headless-react/src/ssr/client-utils.ts

```diff
@@ -1,23 +1,49 @@
 'use client';
 
-import {useEffect, useReducer, useRef} from 'react';
+import {DependencyList, useEffect, useReducer, useRef} from 'react';
+
+/**
+ * Subscriber is a function that takes a single argument, which is another function `listener` that returns `void`. The Subscriber function itself returns another function that can be used to unsubscribe the `listener`.
+ */
+export type Subscriber = (listener: () => void) => () => void;
+
+export type SnapshotGetter<T> = () => T;
+
+/**
+ * Determine if the given list of dependencies has changed.
+ */
+function useHasDepsChanged(deps: DependencyList) {
+  const ref = useRef<DependencyList | null>(null);
+  if (ref.current === null) {
+    ref.current = deps;
+    return false;
+  }
+  if (
+    ref.current.length === deps.length &&
+    !deps.some((dep, i) => !Object.is(ref.current![i], dep))
+  ) {
+    return false;
+  }
+  ref.current = deps;
+  return true;
+}
 
 /**
  * Alternate for `useSyncExternalStore` which runs into infinite loops when hooks are used in `getSnapshot`
  * https://github.com/facebook/react/issues/24529
  */
-export function useSyncMemoizedStore<T>(
-  subscribe: (listener: () => void) => () => void,
-  getSnapshot: () => T
+export function useSyncMemoizedStore<T extends {}>(
+  subscribe: Subscriber,
+  getSnapshot: SnapshotGetter<T>
 ) {
-  const stateRef = useRef(getSnapshot());
+  const snapshot = useRef({});
```

Comment:
Concerning your comment about this being null, it may be possible to make this null by passing it as a generic like so:
```ts
const snapshot = useRef<T | null>(null);
```

If the problem is that `T` may be nullable and we don't want to identify `null` as being "call `getSnapshot` immediately", then the ref could be split into two like so:
```ts
const snapshot = useRef<T | null>(null);
const snapshotIsSet = useRef<boolean>(false);
// ...
if (!snapshotIsSet.current) {
  snapshot.current = getSnapshot();
  snapshotIsSet.current = true;
}
// ...
return snapshot.current as T;
```